Add dropdown to Movie Maker button in editor run bar to access settings

Meorge opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/11541. Supersedes #104954.

This PR converts the Movie Maker button in the editor run bar from a simple toggle to a dropdown button (MenuButton) with a checkbox item to enable or disable Movie Maker mode, and another item to quickly open the Project Settings section for Movie Maker.

https://github.com/user-attachments/assets/52c69d00-4a28-4eeb-bdb7-6ce15a7f2008

The previous PR (#104954) contained a *bit* more functionality than this - it had an output filepath entry field in the dropdown itself - but I found it was finicky and didn't save much time. This is a safer implementation, in my opinion.

Meorge committed:
Add dropdown to Movie Maker button in editor run bar to access settings

Display Movie Maker button as "pressed" when Movie Maker mode is enabled

Fix Movie Maker button colors

Use enum for Movie Maker menu options

Fix Movie Maker button background when switching themes

Apply suggestions from code review

Co-authored-by: Tomasz Chabora <kobewi4e@gmail.com>

Use a theme variation for button colors

Update editor/gui/editor_run_bar.h

Co-authored-by: A Thousand Ships <96648715+AThousandShips@users.noreply.github.com>

RedMser commented:
Looking at your video, it seems like the `MovieWriterButtonNormal` theme style isn't applying anymore.

Meorge commented:
I can double check it later, but I'm pretty sure it is working. I think it's just the custom style I have installed that's much lower contrast than the default - IIRC without the PR code, the appearance is the same when the button's been toggled. Again though I'll have to double check

Meorge commented:
I reverted my editor to the default theme and took screenshots of the editor run bar when Movie Maker mode is enabled:

On `master`:
![CleanShot 2025-05-10 at 10 59 48@2x](https://github.com/user-attachments/assets/91aed92a-e90e-4181-b2e1-577861c478d0)

On my PR:
![CleanShot 2025-05-10 at 10 58 43@2x](https://github.com/user-attachments/assets/36538870-ee01-444c-b50e-0a741a466706)

So the overall highlight effect on the editor run bar is still working, but the Movie Maker button itself no longer has the "pressed" appearance (which I guess makes sense in retrospect, since I removed the toggle and "pressed" state code). I might see if I can emulate the "pressed" appearance via a custom editor style.

Meorge commented:
I've updated my PR to use the "pressed" style on the button when Movie Maker mode is enabled:


https://github.com/user-attachments/assets/97bb0f1e-859a-48d4-ac2b-be78f94f854e

Calinou reviewed (commented):
Tested locally, it works as expected. This is a great usability improvement :slightly_smiling_face: 

I noticed an UI issue though: when you click the icon in the top-right corner, it becomes black until you close the popup:

![image](https://github.com/user-attachments/assets/55daeaa4-3596-47d8-b5c6-efcf62af9f04)

When enabled, clicking the popup then clicking outside will result in the opposite issue (bright icon on bright background):

![image](https://github.com/user-attachments/assets/dea7e5c4-38e1-4d31-99a8-8f36bb627e9b)

Meorge commented:
Thanks for reviewing, and I'm glad you like it! 😄 

I've updated the theming for the button so that it now displays better when on the default theme:

https://github.com/user-attachments/assets/71b01f10-512b-4a53-9151-3b73711417f7

Calinou reviewed (commented):
It looks good now (tested on dark and light themes).

One remaining issue is that if you switch from a light to dark theme while Movie Maker mode is enabled, the icon's color looks wrong:

![image](https://github.com/user-attachments/assets/e74c9282-ebd8-4e56-b2a8-0a769c6846f9)

This also occurs if you switch from dark to light theme while it's active:

![image](https://github.com/user-attachments/assets/f6d521eb-664c-408b-9ea6-d0ab12988373)

If you toggle it off and back on, you get the intended result:

![image](https://github.com/user-attachments/assets/842349b5-084b-40f4-9bdb-9effc0c4c928)

Meorge commented:
Thanks! I tested it and should have fixed it now.

KoBeWi commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -655,13 +683,16 @@ EditorRunBar::EditorRunBar() {
-	write_movie_button = memnew(Button);
+	MenuButton *write_movie_menu_button = memnew(MenuButton);
+	PopupMenu *write_movie_popup = write_movie_menu_button->get_popup();
+	write_movie_popup->add_check_item("Enable Movie Maker Mode", TOGGLE_MOVIE_MAKER);
+	write_movie_popup->add_item("Open Movie Maker Settings...", OPEN_MOVIE_MAKER_SETTINGS);
```
```suggestion
	write_movie_popup->add_check_item(TTRC("Enable Movie Maker Mode"), TOGGLE_MOVIE_MAKER);
	write_movie_popup->add_item(TTRC("Open Movie Maker Settings..."), OPEN_MOVIE_MAKER_SETTINGS);
```

KoBeWi commented on editor/gui/editor_run_bar.h:
```diff
@@ -33,9 +33,11 @@
 #include "editor/editor_run.h"
 #include "editor/export/editor_export.h"
 #include "scene/gui/margin_container.h"
+#include "scene/gui/menu_button.h"
```
This should be in cpp file.

KoBeWi commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -655,13 +683,16 @@ EditorRunBar::EditorRunBar() {
 	main_hbox->add_child(write_movie_panel);
 
-	write_movie_button = memnew(Button);
+	MenuButton *write_movie_menu_button = memnew(MenuButton);
```
Why this isn't `write_movie_button`?

Meorge replied:
Good question... I think it was from me trying to replicate the "Run in Regular/XR Mode" MenuButton code, which works that same way. But since we're always setting `write_movie_button` to the same thing, it's unnecessary here.

KoBeWi commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -95,20 +95,30 @@ void EditorRunBar::_notification(int p_what) {
 			if (is_movie_maker_enabled()) {
 				main_panel->add_theme_style_override(SceneStringName(panel), get_theme_stylebox(SNAME("LaunchPadMovieMode"), EditorStringName(EditorStyles)));
+
+				write_movie_button->begin_bulk_theme_override();
```
You can use a theme variation instead.

Meorge replied:
I haven't done that much, so I'll have to look into how it works. I believe this is how the Movie Maker button styling was done before my PR so at least it's in line with that, but if theme variations would be better for this then I can certainly try to implement them here!

KoBeWi commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -157,6 +167,23 @@ void EditorRunBar::_update_play_buttons() {
 
+void EditorRunBar::_movie_maker_item_pressed(int p_id) {
+	switch (p_id) {
+		case TOGGLE_MOVIE_MAKER: { // Toggle Movie Maker mode.
```
```suggestion
		case TOGGLE_MOVIE_MAKER: {
```
No reason to repeat the enum.

Meorge replied:
Makes sense, will do!

KoBeWi commented on editor/gui/editor_run_bar.h:
```diff
@@ -72,8 +74,13 @@ class EditorRunBar : public MarginContainer {
 	EditorRunNative *run_native = nullptr;
 
+	enum MovieMakerMenuItem {
+		TOGGLE_MOVIE_MAKER,
+		OPEN_MOVIE_MAKER_SETTINGS
```
```suggestion
		MOVIE_MAKER_TOGGLE,
		MOVIE_MAKER_OPEN_SETTINGS
```
Due to how enums work in C++ (the values are available without using the enum name), we usually prefix them to avoid potential conflicts and better organize the values.

Meorge reviewed (commented):
Thanks for the feedback! I'll work on getting that incorporated soon.

Calinou reviewed (approved):
Works great on my end now :slightly_smiling_face: 

One possible improvement is to use a white icon when the calculated background color (according to the current accent color and dark/light theme mode) is dark enough. The black icon is quite hard to distinguish here right now:

![image](https://github.com/user-attachments/assets/96806c76-107c-4e24-bfec-1fb5e640492b)

We have a similar issue in the subresource inspector on light theme (black text on a fairly dark background):

![image](https://github.com/user-attachments/assets/2b0a1d35-26da-461f-8f1e-ae75c2098b74)

It's not critical though, so this can be left to a future PR if it's nontrivial to do.

Repiteo commented:
Thanks!

inhalt120g commented:
As someone who makes lots of movies… When I first saw this feature, I thought it's a great improvement but after using this button for a while, I have to say I really don't like it.

The main problem is that every time I want to enable / disable Movie Maker, I have to click twice in order to do that (first click on the button, then tick / untick the box). The dropdown menu is just slowing things down. It's needlessly adding yet another button that has to be clicked. Before, it was just a simple toggle which was simple and fast.

As is, it's also introducing inconsistencies. Right next to it, there is the Remote Deploy button, which unlike the current Movie Maker button does not offer you to open something like Build / Export settings (and why should it).

All the problems it tries to solve are better solved in other ways, for example in the case of #11541 : keep a File Explorer / Finder / whatever window with the target directory open and after Godot generates the file just rename the file in the file manager as you see fit. (chances are you'll have to open that folder in the file manager later anyway if you want to use the file, and it makes no difference if you input the name in the file manager or Godot)
